src/mps/diskprojmpo.jl

```diff
@@ -0,0 +1,101 @@
+
+"""
+A DiskProjMPO computes and stores the projection of an
+MPO into a basis defined by an MPS, leaving a
+certain number of site indices of the MPO unprojected.
+Which sites are unprojected can be shifted by calling
+the `position!` method.
+
+Drawing of the network represented by a ProjMPO `P(H)`, 
+showing the case of `nsite(P)==2` and `position!(P,psi,4)` 
+for an MPS `psi`:
+
+```
+o--o--o-      -o--o--o--o--o--o <psi|
+|  |  |  |  |  |  |  |  |  |  |
+o--o--o--o--o--o--o--o--o--o--o H
+|  |  |  |  |  |  |  |  |  |  |
+o--o--o-      -o--o--o--o--o--o |psi>
+```
+
+The environment tensors are stored on disk, which is helpful
+for large bond dimensions if they cannot fit in memory.
+"""
+mutable struct DiskProjMPO <: AbstractProjMPO
+  lpos::Int
+  rpos::Int
+  nsite::Int
+  H::MPO
+  LR::DiskVector{ITensor}
+  Lcache::Union{ITensor,Nothing}
+  lposcache::Union{Int,Nothing}
+  Rcache::Union{ITensor,Nothing}
+  rposcache::Union{Int,Nothing}
+end
+
+function DiskProjMPO(H::MPO)
+  return new(
+    0,
+    length(H) + 1,
+    2,
+    H,
+    disk(Vector{ITensor}(undef, length(H))),
+    nothing,
+    nothing,
+    nothing,
+    nothing,
+  )
+end
+
+function disk(pm::ProjMPO)
+  return DiskProjMPO(
+    pm.lpos, pm.rpos, pm.nsite, pm.H, disk(pm.LR), lproj(pm), pm.lpos, rproj(pm), pm.rpos
+  )
+end
+disk(pm::DiskProjMPO) = pm
+
+# Special overload of lproj which uses the cached
+# version of the left projected MPO, and if the
+# cache doesn't exist it loads it from disk.
+function lproj(P::DiskProjMPO)::Union{ITensor,Nothing}
+  (P.lpos <= 0) && return nothing
+  if (P.lpos ≠ P.lposcache) || (P.lpos == 1)
+    # Need to update the cache
+    P.Lcache = P.LR[P.lpos]
+    P.lposcache = P.lpos
```

Comment:
An alternative to this approach (which definitely works fine but requires managing both pos variables) would be to have a "setter" function `setlpos!` which sets `Lcache=nothing` if the new position variable passed to it is different from the current one. Then similarly defining a function that is for retrieving L that sets Lcache to P.LR[P.lpos] if isnothing(Lcache).

Reply:
Yeah good point. I don't particularly like how much of this code uses setting the fields directly as opposed to using functions like `setlpos!`, I was just writing it in a way that required the minimal number of changes to get the new functionality.

Reply:
I'm just going to leave this for now, to not mess too much with the current `ProjMPO` code. But we can add that as an issue for further investigation.